SMB admin backups

danielduhh opened the pull request:
https://github.com/AmericanRedCross/posm/issues/141

create 'backups' SMB share directories: 
  /opt/data/backups/osm
  /opt/data/backups/omk
  /opt/data/backups/fieldpapers

grant admin access to osm/root user for related backup scripts

Commit:
create data backup folders for field papers and omk
create permissions for admin to run api-db backup as postgres user

Commit:
Merge branch 'master' into admin-backups

* master:
  Document where to change wireless auth
  Use MM tiles for atlas extent browsing
  Disable FP email confirmation

Commit:
Merge branch 'master' into admin-backups

* master:
  Allow batch uploading by copying over SMB

Commit:
use public share via samba

Commit:
use /opt/data/backups directory

Commit:
update /backups chown & chmod

Commit:
grant access for root_fp-production-db-backup.sh to run as root

Commit:
remove reference to feature branch

hallahan commented on kickstart/scripts/samba-deploy.sh:
```diff
@@ -11,6 +11,18 @@ deploy_samba_ubuntu() {
 
+  mkdir -p /opt/data/backups/omk
+  chown nobody:nogroup /opt/data/backups/omk
+  chmod 777 /opt/data/backups/omk
```
Do we want to have nobody own it and give full rights to the whole system? Is there a reason behind this?

Samba needs this?

danielduhh replied:
great question for @mojodna; I gave each directory the same rights as https://github.com/AmericanRedCross/posm-build/pull/36/commits/13ec0a8c77276d955e365207ece0578accbc70d6#diff-87896bd638e154ba9e87ed58def78095R10

sethfitz replied:
Not here. The FP batch upload folder and public are `777` because they're intended to be read/written by everyone.

`/opt/data/backups/omk` should be owned by `admin:admin` and be `755` so that guests can read it but not delete directories (they'll be reading as `nobody:nogroup`, hence the `55` / aka `rx`).

@danielduhh can you move the creation of this directory into `admin-deploy.sh`, as it's really posm-admin-related?

hallahan committed:
make samba-deploy.sh executable

hallahan committed:
added omk to samba config

hallahan commented:
I'm having issues connecting to SMB via Mac Finder at all now... I added omk to the `smb.conf`, and then restarted samba. At first, it was working and showing `fieldpapers` and `public`. Now, nothing shows up.

I then changed `smb.conf` back and restarted the service. It still is not working. Not sure whats up.

Could you give it a try @mojodna ?

https://github.com/AmericanRedCross/posm-build/pull/45/commits/187329b580f7938bcfcfb2d61dd10442299bfdaf

cc/ @danielduhh @mojodna

Commit:
add fp & osm to samba cfg

https://github.com/AmericanRedCross/posm-build/pull/45
cc/ @hallahan

hallahan commented:
Odd, pulled in @danielduhh changes and redeployed samba. Still not working, though POSM does pop up after I sit in the networks finder view for a while. But, then it does a _Connection Failed_.

<img width="770" alt="screenshot 2016-05-23 16 16 51" src="https://cloud.githubusercontent.com/assets/556367/15487391/e36c5ce4-2101-11e6-8abd-80dc9d0cd048.png">

<img width="771" alt="screenshot 2016-05-23 16 17 02" src="https://cloud.githubusercontent.com/assets/556367/15487397/e75a69fe-2101-11e6-9820-5134aad0bc42.png">

hallahan commented:
If we go into finder and explicitly connect to the server (Cmd K) `smb://posm`, it works. There is some sort of samba problem that is causing the samba service to crash.

![screenshot 2016-05-24 13 20 40](https://cloud.githubusercontent.com/assets/556367/15518562/5970a002-21b2-11e6-9079-d4a8cd5f086a.png)

I tried commenting out interfaces, bind interfaces, and server role in `/etc/samba/smb.conf` to the same result:

```
[global]
  workgroup = POSM
  server string = POSM
  dns proxy = no
  security = user
  #interfaces = wlan0
  #bind interfaces only = true
  guest account = nobody

  # Ubuntu defaults
  log file = /var/log/samba/log.%m
  max log size = 1000
  syslog = 0
  panic action = /usr/share/samba/panic-action %d
  #server role = standalone server
  passdb backend = tdbsam
  obey pam restrictions = yes
  unix password sync = yes
  passwd program = /usr/bin/passwd %u
  passwd chat = *Enter\snew\s*\spassword:* %n\n *Retype\snew\s*\spassword:* %n\n *password\supdated\ssuccessfully* .
  pam password change = yes
  map to guest = bad user
  usershare allow guests = yes
```

Looking at `/var/log/samba/log.smbd`, we see:

```
[2016/05/24 18:15:22.541318,  0] ../source3/lib/interface.c:543(load_interfaces)
  WARNING: no network interfaces found
[2016/05/24 18:15:22.597882,  0] ../lib/util/become_daemon.c:124(daemon_ready)
  STATUS=daemon 'smbd' finished starting up and ready to serve connections
[2016/05/24 18:15:22.599276,  0] ../source3/smbd/server.c:877(open_sockets_smbd)
  open_sockets_smbd: No sockets available to bind to.
[2016/05/24 18:15:22.599677,  0] ../source3/lib/util.c:789(smb_panic_s3)
  PANIC (pid 630): open_sockets_smbd() failed
[2016/05/24 18:15:22.602028,  0] ../source3/lib/util.c:900(log_stack_trace)
  BACKTRACE: 9 stack frames:
   #0 /usr/lib/x86_64-linux-gnu/samba/libsmbregistry.so.0(log_stack_trace+0x1a) [0x7fdd187cf16a]
   #1 /usr/lib/x86_64-linux-gnu/samba/libsmbregistry.so.0(smb_panic_s3+0x20) [0x7fdd187cf240]
   #2 /usr/lib/x86_64-linux-gnu/libsamba-util.so.0(smb_panic+0x2f) [0x7fdd195468ef]
   #3 /usr/lib/x86_64-linux-gnu/samba/libsmbd-base.so.0(+0x14caa5) [0x7fdd19149aa5]
   #4 /usr/lib/x86_64-linux-gnu/samba/libsmbd-base.so.0(+0x14cd81) [0x7fdd19149d81]
   #5 /usr/lib/x86_64-linux-gnu/samba/libsmbd-shim.so.0(exit_server+0x12) [0x7fdd17027c82]
   #6 smbd(main+0x1133) [0x55f1884ee243]
   #7 /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf5) [0x7fdd15d15ec5]
   #8 smbd(+0x7a96) [0x55f1884eea96]
[2016/05/24 18:15:22.602191,  0] ../source3/lib/util.c:801(smb_panic_s3)
  smb_panic(): calling panic action [/usr/share/samba/panic-action 630]
[2016/05/24 18:15:22.609929,  0] ../source3/lib/util.c:809(smb_panic_s3)
  smb_panic(): action returned status 0
[2016/05/24 18:15:22.610033,  0] ../source3/lib/dumpcore.c:318(dump_core)
  dumping core in /var/log/samba/cores/smbd
[2016/05/24 18:15:22.806440,  0] ../source3/lib/interface.c:543(load_interfaces)
  WARNING: no network interfaces found
[2016/05/24 18:15:22.820333,  0] ../lib/util/become_daemon.c:124(daemon_ready)
  STATUS=daemon 'smbd' finished starting up and ready to serve connections
[2016/05/24 18:15:22.821875,  0] ../source3/smbd/server.c:877(open_sockets_smbd)
  open_sockets_smbd: No sockets available to bind to.
[2016/05/24 18:15:22.822260,  0] ../source3/lib/util.c:789(smb_panic_s3)
  PANIC (pid 753): open_sockets_smbd() failed
```

Ok, it's not able to connect to a network interface?

hallahan commented:
I tried commenting out

```
passdb backend = tdbsam
```

Now, after restarting, POSM is showing up in networks, but it takes about a minute for it to pop up. I'm not seeing any errors in the log, however. This may resolve it?

hallahan commented:
No, this doesn't change the problem. Sometimes POSM does show up in the network, but usually not. 

Looking in `/var/log/samba/log.%m`, we are still getting this error, even though I commented out `server role`:

```
[2016/05/24 20:43:35.731781,  0] ../lib/util/become_daemon.c:111(exit_daemon)
  STATUS=daemon failed to start: Samba detected misconfigured 'server role' and exited. Check logs for details, error code 22
```

We're not getting any errors now in `log.smbd` now.

How shall we proceed @mojodna ? This PR is sorta off topic to this problem, I can merge as is. Or, should I commit the config params I commented out?

hallahan commented:
Looks like @mojodna fixed the Mac Finder problem via https://github.com/AmericanRedCross/posm-build/pull/46

sethfitz commented on kickstart/etc/smb.conf:
```diff
@@ -39,3 +39,30 @@
+   guest ok = yes
+   create mask = 0777
+   directory mask = 0777
+   writable = yes
```
Drop these lines from each of the shares since we can inherit filesystem permissions:

```
   create mask = 0777
   directory mask = 0777
   writable = yes
```

sethfitz commented on kickstart/etc/smb.conf:
```diff
@@ -39,3 +39,30 @@
+
+[omk]
+   comment = OpenMapKit Backups
+   path = /opt/data/backups/omk
```
Consolidate these 3 shares into a single `[backup]` share exposing `path = /opt/data/backups`? That way we can add more backups via posm-admin or whatever and they'll "magically" show up shared.

sethfitz commented on kickstart/scripts/admin-deploy.sh:
```diff
@@ -45,13 +45,15 @@ deploy_posm_admin() {
   chown osm:osm "$dst/posm-admin/scripts/osm_api-db-init.sh"
   chown osm:osm "$dst/posm-admin/scripts/osm_api-db-populate.sh"
   chown osm:osm "$dst/posm-admin/scripts/osm_render-db-api2pbf.sh"
   chown gis:gis "$dst/posm-admin/scripts/gis_render-db-pbf2render.sh"
```
FWIW, these don't actually need to be owned by other users, they just need to be executable by them (`a+rx` or `__5`).

sethfitz replied:
(To be clear, the action here is to `chmod -R 755 "$dst/posm-admin/scripts` and remove the `chown` calls.)

sethfitz commented on kickstart/scripts/samba-deploy.sh:
```diff
@@ -11,6 +11,18 @@ deploy_samba_ubuntu() {
+  mkdir -p /opt/data/backups/fieldpapers
+  chown nobody:nogroup /opt/data/backups/osm
+  chmod 777 /opt/data/backups/osm
+
```
Actually, rather than moving these to `admin-deploy.sh`, move them to each of the components that they belong to, i.e. `/opt/data/backups/fieldpapers` to `fieldpapers-deploy.sh`. We're then set up to add appropriate cron jobs and allow each component to manage its (automated) backup strategy.

sethfitz commented:
@danielduhh can you address the stuff I added as line notes before we merge?

danielduhh commented:
@mojodna Yeah, will start tomorrow

Commit:
Merge branch 'master' into admin-backups

* master:
  Prevent Samba from listening on :139
  Disable netbios
  Remove Samba server role
  Send an atlas payload to posm-admin when created

Commit:
remove osm/omk/fp backups from smb config & consolidate to /backups
create backup folders in the deployment components they belong to

https://github.com/AmericanRedCross/posm-build/pull/45

sethfitz commented on kickstart/scripts/fieldpapers-deploy.sh:
```diff
@@ -70,6 +70,11 @@ deploy_fp_web() {
 
+  # create backup directory
+  mkdir -p /opt/data/backups/fieldpapers
+  chown nobody:nogroup /opt/data/backups/fieldpapers
```
`fp:fp`

sethfitz commented on kickstart/scripts/fieldpapers-deploy.sh:
```diff
@@ -70,6 +70,11 @@ deploy_fp_web() {
+  # create backup directory
+  mkdir -p /opt/data/backups/fieldpapers
+  chown nobody:nogroup /opt/data/backups/fieldpapers
+  chmod 777 /opt/data/backups/fieldpapers
```
`644` so that stuff can't be deleted w/o being the owner or `root`.

sethfitz commented on kickstart/scripts/omk-deploy.sh:
```diff
@@ -28,6 +28,11 @@ deploy_omk_server() {
 
+  # create backup directory
+  mkdir -p /opt/data/backups/omk
+  chown nobody:nogroup /opt/data/backups/omk
```
`omk:omk`

sethfitz commented on kickstart/scripts/omk-deploy.sh:
```diff
@@ -28,6 +28,11 @@ deploy_omk_server() {
+  # create backup directory
+  mkdir -p /opt/data/backups/omk
+  chown nobody:nogroup /opt/data/backups/omk
+  chmod 777 /opt/data/backups/omk
```
`644`

sethfitz commented on kickstart/scripts/osm-deploy.sh:
```diff
@@ -89,6 +89,11 @@ deploy_osm_rails() {
 
+  # create backup directory
+  mkdir -p /opt/data/backups/osm
+  chown nobody:nogroup /opt/data/backups/osm
```
`osm:osm`

sethfitz commented on kickstart/scripts/osm-deploy.sh:
```diff
@@ -89,6 +89,11 @@ deploy_osm_rails() {
+  # create backup directory
+  mkdir -p /opt/data/backups/osm
+  chown nobody:nogroup /opt/data/backups/osm
+  chmod 777 /opt/data/backups/osm
```
`644`

sethfitz commented on kickstart/scripts/samba-deploy.sh:
```diff
@@ -11,17 +11,9 @@ deploy_samba_ubuntu() {
-  chmod 777 /opt/data/backups/osm
+  mkdir -p /opt/data/backups
+  chown nobody:nogroup /opt/data/backups
+  chmod 777 /opt/data/backups
```
`644`

sethfitz commented on kickstart/scripts/samba-deploy.sh:
```diff
@@ -11,17 +11,9 @@ deploy_samba_ubuntu() {
-  chown nobody:nogroup /opt/data/backups/osm
-  chmod 777 /opt/data/backups/osm
+  mkdir -p /opt/data/backups
+  chown nobody:nogroup /opt/data/backups
```
Delete this so it's owned by `root:root`.

danielduhh replied:
remove it from samba deploy or just update the change owner to `root:root` ?  

```
mkdir -p /opt/data/backups
chown root:root /opt/data/backups
chmod 644 /opt/data/backups
```

sethfitz replied:
Remove it. Effectively equivalent.

``` bash
mkdir -p /opt/data/backups
chmod 644 /opt/data/backups
```

sethfitz replied:
I'll get this.

sethfitz commented:
@danielduhh minor security permissions / ownership tweaks needed, otherwise good to go.

Commit:
change component backups access to 644 to prevent deletion from non root/owner users

sethfitz commented:
Awesome.
